pkg/tfbridge/x/token_test.go

```diff
@@ -472,46 +473,333 @@ func TestAliasing(t *testing.T) {
+	require.NoError(t, err)
+
+	assert.True(t, *info.Resources["pkg_r1"].Fields["f1"].MaxItemsOne)
+	assert.False(t, *info.Resources["pkg_r1"].Fields["f2"].MaxItemsOne)
+	assert.Equal(t, expected, string(metadata.Marshal()))
+
+	// Apply metadata back into the provider again, making sure there isn't a diff
+	err = AutoAliasing(info, metadata)
+	require.NoError(t, err)
+
+	assert.True(t, *info.Resources["pkg_r1"].Fields["f1"].MaxItemsOne)
+	assert.False(t, *info.Resources["pkg_r1"].Fields["f2"].MaxItemsOne)
+	assert.Equal(t, expected, string(metadata.Marshal()))
+
+	// Validate that overrides work
+
+	info = provider(true, false)
+	info.Resources["pkg_r1"].Fields = map[string]*tfbridge.SchemaInfo{
+		"f1": {MaxItemsOne: tfbridge.False()},
+	}
+
+	err = AutoAliasing(info, metadata)
+	require.NoError(t, err)
+	assert.False(t, *info.Resources["pkg_r1"].Fields["f1"].MaxItemsOne)
+	assert.False(t, *info.Resources["pkg_r1"].Fields["f2"].MaxItemsOne)
+	assert.Equal(t, `{
+    "auto-aliasing": {
+        "resources": {
+            "pkg_r1": {
+                "current": "pkg:index/r1:R1",
+                "fields": {
+                    "f1": {
+                        "maxItemOne": false
+                    },
+                    "f2": {
+                        "maxItemOne": false
+                    }
+                }
+            }
+        },
+        "datasources": {}
+    }
+}`, string(metadata.Marshal()))
+}
+
+func TestMaxItemsOneAliasingExpiring(t *testing.T) {
+	provider := func(f1, f2 bool) *tfbridge.ProviderInfo {
+		prov := &tfbridge.ProviderInfo{
+			P: Provider{
+				resources: map[string]shim.SchemaMap{
+					"pkg_r1": newSchemaMap(map[string]shim.Schema{
+						"f1": Schema{MaxItemsOne: f1},
+						"f2": Schema{MaxItemsOne: f2},
+					}),
+				},
+			},
+		}
+		err := ComputeDefaults(prov, TokensSingleModule("pkg_", "index", MakeStandardToken("pkg")))
+		require.NoError(t, err)
+		return prov
+	}
+	info := provider(true, false)
+	metadata, err := metadata.New(nil)
+	require.NoError(t, err)
+
+	// Save current state into metadata
+	err = AutoAliasing(info, metadata)
+	require.NoError(t, err)
+
+	v := string(metadata.Marshal())
+	expected := `{
+    "auto-aliasing": {
+        "resources": {
+            "pkg_r1": {
+                "current": "pkg:index/r1:R1",
+                "fields": {
+                    "f1": {
+                        "maxItemOne": true
+                    },
+                    "f2": {
+                        "maxItemOne": false
+                    }
+                }
+            }
+        },
+        "datasources": {}
+    }
+}`
+	assert.Equal(t, expected, v)
+
+	info = provider(false, true)
+
+	// Apply metadata back into the provider
+	info.Version = "1.0.0" // New major version
+	err = AutoAliasing(info, metadata)
+	require.NoError(t, err)
+
+	assert.Nil(t, info.Resources["pkg_r1"].Fields["f1"].MaxItemsOne)
+	assert.Nil(t, info.Resources["pkg_r1"].Fields["f2"].MaxItemsOne)
```

Comment:
Could you talk me through this one? 

```info = provider(false, true)```

This makes a provider with `info.Resources["pkg_r1"].Fields["f2"].MaxItemsOne == true` before the auto-alias pass. Why does it become nil after the AutoAlias pass? Weren't we supposed to discard history and leave as-is on a new major version?

Reply:
In this test, we 
1. create a `provider(true, false)`
2. Apply blank metadata on that provider, baking the provider state into `metadata`.
3. Create a new provider with a different `.P` value for `MaxItemsOne` and a new version.
4. We apply the previous `metadata` to the new provider, but `AutoAliasing` discards the old information. Since `AutoAliasing` is applying without previous data, it remembers but doesn't set `info`. (the nil check)

I think your confusion is because `provider(false, true)` does not make a provider with `info.Resources["pkg_r1"].Fields["f2"].MaxItemsOne == true`. It makes a provider with `info.P.resources["pkg_r1"].Fields["f2"].MaxItemsOne == true`, which is different.

Reply:
Got it. Correct but confusing. The new version wipes the "user overrides" to nil, which is the info.Resources while info.P.Resources is what the provider wants.